docs/develop/plugins/wasm/best_practices.md

```diff
@@ -0,0 +1,738 @@
+journalctl -u signalk -f | grep wasm
+```
+
+### Common Issues
+
+**Issue**: Plugin doesn't load
+**Solution**: Check `wasmManifest` path in package.json
+
+**Issue**: Capability errors
+**Solution**: Ensure required capabilities declared in package.json
+
+**Issue**: Crashes on start
+**Solution**: Check server logs for error details
+
+## Migration from Node.js
+
+### 1. Assess Compatibility
+
+Check if your plugin:
+
+- ✅ Processes deltas
+- ✅ Reads/writes configuration
+- ✅ Uses data model APIs
+- ✅ Registers REST endpoints
+- ❌ Uses serial ports (wait for Phase 3)
+- ✅ Makes network requests
+
+### 2. Port Logic to Rust
+
+Convert TypeScript/JavaScript logic to Rust:
+
+**Before (Node.js):**
+
+```javascript
+plugin.start = function (config) {
+  app.handleMessage('my-plugin', {
+    updates: [{ values: [{ path: 'foo', value: 'bar' }] }]
+  })
+}
+```
+
+**After (WASM/Rust):**
+
+```rust
+fn start(config_ptr: *const u8, config_len: usize) -> i32 {
+    let delta = json!({
+        "updates": [{ "values": [{ "path": "foo", "value": "bar" }] }]
+    });
+    sk_emit_delta(&delta.to_string());
+    0
+}
+```
+
+### 3. Migrate Data
+
+Use migration helper to copy existing data to VFS:
+
+```rust
+fn first_run_migration() {
+    // Server provides migration API
+    // Copies files from ~/.signalk/plugin-config-data/{id}/
+    // to ~/.signalk/plugin-config-data/{id}/vfs/data/
+}
+```
+
+## Example Plugins
+
+### Hello World
+
+Minimal example that emits a delta on start - see the [AssemblyScript guide](assemblyscript.md) Step 3.
+
+### Data Logger
+
+Logs vessel data to VFS:
+
+```rust
+use std::fs::OpenOptions;
+use std::io::Write;
+
+fn log_data(path: &str, value: &str) {
+    let mut file = OpenOptions::new()
+        .create(true)
+        .append(true)
+        .open("/data/log.csv")
+        .unwrap();
+
+    let timestamp = get_current_timestamp();
+    writeln!(file, "{},{},{}", timestamp, path, value).unwrap();
+}
+```
+
+### Derived Data Calculator
+
+Calculates derived values from sensor data:
+
+```rust
+fn calculate_true_wind() {
+    let aws = get_self_path("environment.wind.speedApparent");
+    let awa = get_self_path("environment.wind.angleApparent");
+    let sog = get_self_path("navigation.speedOverGround");
+    let cog = get_self_path("navigation.courseOverGroundTrue");
+
+    if let (Some(aws), Some(awa), Some(sog), Some(cog)) = (aws, awa, sog, cog) {
+        let (tws, twa) = calculate_true_wind_values(aws, awa, sog, cog);
+        emit_true_wind_delta(tws, twa);
+    }
+}
+```
+
+## Advanced Features
+
+### Static File Serving
+
+Plugins can serve HTML, CSS, JavaScript and other static files:
+
+**Structure:**
+
+```
+@signalk/my-plugin/
+├── public/           # Automatically served at /plugins/my-plugin/
+│   ├── index.html
+│   ├── style.css
+│   └── app.js
+├── plugin.wasm
+└── package.json
+```
+
+**Access:** `http://localhost:3000/plugins/my-plugin/` serves `public/index.html`
+
+### Privileged Operations (Optional)
```

Comment:
I don't think this feature is worth the extra hassle and sandbox escape vector. It seems like you wanted to do the log access plugin in WASM, so that is included as a feature somewhat arbitrarily. I have only heard of one plugin that wants log file access, so not generally useful.

We can add this later, if it turns out to be needed. If it is added I would want the user to have explicit control over the privilege to execute additional commands.